Resolve 'hass undefined' error in conditional-card

iantrich opened the pull request:
Fix for https://github.com/home-assistant/home-assistant-polymer/issues/1905

iantrich committed:
Resolve 'hass undefined' error in conditional-card

balloob commented on src/panels/lovelace/cards/hui-conditional-card.ts:
```diff
@@ -32,7 +32,9 @@ class HuiConditionalCard extends HTMLElement implements LovelaceCard {
     if (this.card) {
       this.appendChild(this.card);
-      this.card.hass = this.hass;
+      if (this.hass) {
```
this is weird. There is no `this.hass` to get, I only see a setter, no getter.  Neither is the setter storing the value ot `this._hass`.

balloob commented on src/panels/lovelace/cards/hui-conditional-card.ts:
```diff
@@ -32,7 +32,9 @@ class HuiConditionalCard extends HTMLElement implements LovelaceCard {
     this.card = createCardElement(config.card);
     if (this.card) {
       this.appendChild(this.card);
```
this card will break if we set config twice, it will have 2 cards attached to it. We need to clean that up.

balloob commented on src/panels/lovelace/cards/hui-conditional-card.ts:
```diff
@@ -32,7 +32,9 @@ class HuiConditionalCard extends HTMLElement implements LovelaceCard {
       this.appendChild(this.card);
-      this.card.hass = this.hass;
+      if (this.hass) {
+        this.card.hass = this.hass;
```
This is not correct, there is no need to set it if it's going to be invisible. We should do `this.hass = this._hass` and trigger our hass setting logic.

iantrich replied:
I think this is the main takeaway, correct? Like we are doing for the stack cards, have a private hass and setter that is applied once and only if visible.

balloob replied:
yeah.

zsarnett commented:
Make sure to set parameter types and return types for this card in the PR.

iantrich commented:
You think we should be doing that on a hot fix? Would rather just address the bug

zsarnett commented:
True. I kind of forgot this was needed as a hotfix, my b

iantrich committed:
Address review comments

I feel like I may have gone overboard for a hotfix but this is passing all the test I have thrown at it.

zsarnett commented on src/panels/lovelace/cards/hui-conditional-card.ts:
```diff
@@ -49,15 +44,43 @@ class HuiConditionalCard extends HTMLElement implements LovelaceCard {
+      return html``;
+    }
+
+    return html`
```
Should be able to just return the card

iantrich replied:
```
return html`${this._card}`
```
I went ahead and tested that with an undefined and you are correct. Thanks.

zsarnett replied:
I was talking about `return this._card;`

iantrich replied:
`this._card` is not a `TemplateResult`

iantrich replied:
Just double checked and TS does not like that

balloob replied:
The `render` function needs to return `TemplateResult`. All other functions are placing things in a template and thus can be elements, strings etc. So we should adjust the method signature.

zsarnett commented on src/panels/lovelace/cards/hui-conditional-card.ts:
```diff
@@ -49,15 +44,43 @@ class HuiConditionalCard extends HTMLElement implements LovelaceCard {
+      !config.conditions ||
+      !Array.isArray(config.conditions) ||
+      !config.conditions.every(
+        (c) =>
```
This is just a personal preference, not sure if best practice but using the singular version of what the array is so in this card `condition`

iantrich replied:
agreed

zsarnett commented on src/panels/lovelace/cards/hui-conditional-card.ts:
```diff
@@ -49,15 +44,43 @@ class HuiConditionalCard extends HTMLElement implements LovelaceCard {
+      !Array.isArray(config.conditions) ||
+      !config.conditions.every(
+        (c) =>
+          c.hasOwnProperty("entity") &&
```
What is the reason behind using `hasOwnProperty`?

iantrich replied:
TypeScript linter doesn't like the check c.entity `'string' is not assignable to type 'boolean'`. Debated using `!!c.entity` but `hasOwnProperty` felt more proper. 🤷‍♂️

zsarnett replied:
Idk the proper syntax here. Sorry. I was just wondering. You could as type cast but I think what you have or `!!` is better

balloob replied:
alternative is to use `('entity' in c)`.

iantrich committed:
Addressed review comments

iantrich committed:
Add return type to `get properties`

balloob commented on src/panels/lovelace/cards/hui-conditional-card.ts:
```diff
@@ -1,70 +1,88 @@
+  conditions: Condition[];
+}
+
+class HuiConditionalCard extends LitElement implements LovelaceCard {
```
I don't understand, why are you converting this to `LitElement` ? That's not needed ?

iantrich replied:
I thought that was something we were doing?

balloob replied:
we are converting from Polymer Element to Lit Element. This element was an HTML Element, which is very basic with no templating etc, but that's fine because the card only shows/hides an element based on a condition.

iantrich replied:
Is there a good rule of thumb on when we should use Lit vs plain HTML?

iantrich commented:
> I've opened PR #1927 that will fix #1905 without converting it to Lit Element. If we can do it with just an HTML element, we should. It's always faster.

Seeing this comment just now. Couldn't we do things like the stacks with just HTML elements?

balloob commented:
99% of the time, use Lit. Only if the component is extremely simple and not going to use any templating or ShadowDOM (like scoped CSS), use HTMLElement. So in case of the conditional card, it just showes/hides a single element, we could stick with HTMLElement.
